File: medical-prep-center/backend/app/api/academic_content/materials.py
from fastapi import APIRouter, HTTPException, status, Depends, Query, Path
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional

# Импорты базы данных и моделей
from app.database import get_db
from app.database.models.academic import MaterialFileType

# Импорты сервисов
from app.services.content.topic_service import (
    get_topic_with_details_db,
    add_material_file_db,
    search_materials_db
)
from app.services.content.subject_service import (
    get_subject_materials_db, 
    get_materials_statistics_db
)
from app.services.content.section_service import get_section_with_materials_db

# Импорты схем
from app.schemas.assessment.materials import (
    MaterialsResponseModel,
    SectionResponse,
    TopicDetailResponse,
    MaterialFileCreate,
    SearchResult,
    MaterialsStatistics
)

# Создание роутера
router = APIRouter()

# ===========================================
# ПОЛУЧЕНИЕ МАТЕРИАЛОВ ПО ПРЕДМЕТУ
# ===========================================

@router.get("/{subject_name}", 
            response_model=MaterialsResponseModel,
            summary="Получить материалы предмета",
            description="Получение всех материалов по предмету (chemistry/biology)")
async def get_materials_by_subject(
    subject_name: str = Path(..., description="Название предмета"),
    db: AsyncSession = Depends(get_db)
) -> MaterialsResponseModel:
    """
    Получение всех материалов по предмету
    
    - **subject_name**: Название предмета (chemistry или biology)
    
    Возвращает структуру материалов со всеми разделами, темами и файлами
    """
    try:
        # Валидация названия предмета
        valid_subjects = ['chemistry', 'biology', 'химия', 'биология']
        if subject_name.lower() not in valid_subjects:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Недопустимое название предмета. Используйте: chemistry или biology"
            )
        
        # Получение данных
        subject_data = await get_subject_materials_db(db, subject_name)
        
        # Формирование ответа в нужном формате
        materials_data = {
            subject_name.lower(): subject_data
        }
        
        return MaterialsResponseModel(materials=materials_data)
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Внутренняя ошибка сервера: {str(e)}"
        )

# ===========================================
# ПОЛУЧЕНИЕ МАТЕРИАЛОВ РАЗДЕЛА
# ===========================================

@router.get("/section/{section_id}", 
            response_model=SectionResponse,
            summary="Получить материалы раздела",
            description="Получение всех материалов конкретного раздела")
async def get_section_materials(
    section_id: int,
    db: AsyncSession = Depends(get_db)
) -> SectionResponse:
    """
    Получение материалов конкретного раздела
    
    - **section_id**: ID раздела
    
    Возвращает все файлы, темы и информацию о разделе
    """
    try:
        if section_id <= 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="ID раздела должен быть положительным числом"
            )
        
        section_data = await get_section_with_materials_db(db, section_id)
        return SectionResponse(**section_data)
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка получения материалов раздела: {str(e)}"
        )

# ===========================================
# ПОЛУЧЕНИЕ ДЕТАЛЕЙ ТЕМЫ
# ===========================================

@router.get("/topic/{topic_id}", 
            response_model=TopicDetailResponse,
            summary="Получить детали темы",
            description="Получение подробной информации о теме")
async def get_topic_details(
    topic_id: int,
    db: AsyncSession = Depends(get_db)
) -> TopicDetailResponse:
    """
    Получение деталей конкретной темы
    
    - **topic_id**: ID темы
    
    Возвращает все детали темы включая домашнее задание, видео и тесты
    """
    try:
        if topic_id <= 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="ID темы должен быть положительным числом"
            )
        
        topic_data = await get_topic_with_details_db(db, topic_id)
        return TopicDetailResponse(**topic_data)
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка получения темы: {str(e)}"
        )

# ===========================================
# ДОБАВЛЕНИЕ ФАЙЛА К РАЗДЕЛУ
# ===========================================

@router.post("/section/{section_id}/files",
             summary="Добавить файл к разделу",
             description="Добавление нового файла материала к разделу")
async def add_section_file(
    section_id: int,
    file_data: MaterialFileCreate,
    db: AsyncSession = Depends(get_db)
):
    """
    Добавление файла к разделу
    
    - **section_id**: ID раздела
    - **file_data**: Данные файла (название, автор, размер, тип)
    
    Возвращает подтверждение создания файла
    """
    try:
        if section_id <= 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="ID раздела должен быть положительным числом"
            )
        
        # Валидация типа файла
        if file_data.file_type not in ['book', 'test_book']:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Тип файла должен быть 'book' или 'test_book'"
            )
        
        # Создание файла
        material_file = await add_material_file_db(
            db=db, 
            section_id=section_id,
            file_type=MaterialFileType(file_data.file_type),
            title=file_data.title,
            author=file_data.author,
            file_size=file_data.size,
            file_format=file_data.format,
            download_url=file_data.url
        )
        
        return {
            "status": "success", 
            "message": "Файл успешно добавлен",
            "file_id": material_file.file_id,
            "data": {
                "id": material_file.file_id,
                "title": material_file.title,
                "author": material_file.author,
                "size": material_file.file_size,
                "format": material_file.file_format
            }
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка добавления файла: {str(e)}"
        )

# ===========================================
# ПОИСК ПО МАТЕРИАЛАМ
# ===========================================

@router.get("/search", 
            response_model=List[SearchResult],
            summary="Поиск материалов",
            description="Поиск по всем материалам с фильтрацией")
async def search_materials(
    query: str = Query(..., min_length=2, max_length=100, description="Поисковый запрос"),
    subject_filter: Optional[str] = Query(None, description="Фильтр по предмету"),
    section_filter: Optional[int] = Query(None, description="Фильтр по разделу"),
    limit: Optional[int] = Query(20, le=100, description="Лимит результатов"),
    db: AsyncSession = Depends(get_db)
) -> List[SearchResult]:
    """
    Поиск по материалам
    
    - **query**: Поисковый запрос (минимум 2 символа)
    - **subject_filter**: Фильтр по предмету (опционально)
    - **section_filter**: Фильтр по разделу (опционально)
    - **limit**: Максимальное количество результатов (по умолчанию 20)
    
    Возвращает список найденных материалов
    """
    try:
        # Валидация параметров
        if not query or len(query.strip()) < 2:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Поисковый запрос должен содержать минимум 2 символа"
            )
        
        if subject_filter and subject_filter.lower() not in ['chemistry', 'biology', 'химия', 'биология']:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Недопустимый фильтр по предмету"
            )
        
        if section_filter and section_filter <= 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="ID раздела должен быть положительным числом"
            )
        
        # Выполнение поиска
        results = await search_materials_db(
            db=db, 
            query=query.strip(), 
            subject_filter=subject_filter, 
            section_filter=section_filter
        )
        
        # Ограничение результатов
        limited_results = results[:limit] if limit else results
        
        return [SearchResult(**result) for result in limited_results]
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка поиска: {str(e)}"
        )

# ===========================================
# СТАТИСТИКА МАТЕРИАЛОВ
# ===========================================

@router.get("/statistics", 
            response_model=MaterialsStatistics,
            summary="Статистика материалов",
            description="Получение общей статистики по материалам")
async def get_statistics(
    db: AsyncSession = Depends(get_db)
) -> MaterialsStatistics:
    """
    Получение статистики материалов
    
    Возвращает общую информацию о количестве предметов, разделов, тем и файлов
    """
    try:
        stats = await get_materials_statistics_db(db)
        return MaterialsStatistics(**stats)
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка получения статистики: {str(e)}"
        )

# ===========================================
# ПРОВЕРКА ЗДОРОВЬЯ API
# ===========================================

@router.get("/health",
            summary="Проверка здоровья API",
            description="Простая проверка работоспособности API материалов")
async def health_check():
    """
    Проверка здоровья API материалов
    """
    return {
        "status": "healthy",
        "service": "materials-api",
        "timestamp": "2024-01-01T00:00:00Z",
        "version": "1.0.0"
    }

File: medical-prep-center/frontend/src/services/api/materials.ts
// ========== КОНСТАНТЫ ==========

export const SEARCH_RESULT_TYPES = {
  TOPIC: 'topic',
  SECTION: 'section', 
  FILE: 'file'
} as const;

export const FILE_TYPES = {
  BOOK: 'book',
  TEST_BOOK: 'test_book'
} as const;

export const THEMES = {
  LIGHT: 'light',
  DARK: 'dark'
} as const;

export const LANGUAGES = {
  RU: 'ru',
  EN: 'en'
} as const;

// ========== ТИПЫ ==========

export type SearchResultType = typeof SEARCH_RESULT_TYPES[keyof typeof SEARCH_RESULT_TYPES];
export type FileType = typeof FILE_TYPES[keyof typeof FILE_TYPES];
export type Theme = typeof THEMES[keyof typeof THEMES];
export type Language = typeof LANGUAGES[keyof typeof LANGUAGES];

export interface DownloadableFile {
  id: number;
  title: string;
  author: string;
  size: string;
  format: string;
  download_url?: string;
}

export interface Topic {
  id: number;
  title: string;
  homework: string[];
  videoUrl: string;
  testId: number;
  block_name?: string;
}

export interface Section {
  id: number;
  name: string;
  description?: string;
  books: DownloadableFile[];
  testBooks: DownloadableFile[];
  topics: Topic[];
}

export interface SubjectData {
  modules: Section[];
}

export interface MaterialsData {
  [subjectName: string]: SubjectData;
}

export interface SearchResult {
  type: SearchResultType;
  id: number;
  title: string;
  description?: string;
  subject_name: string;
  section_name?: string;
}

export interface MaterialsStatistics {
  total_subjects: number;
  total_sections: number;
  total_topics: number;
  total_files: number;
  subjects_breakdown: Record<string, number>;
}

export interface APIResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface APIErrorInterface {
  status?: number;
  code?: string;
  message: string;
}

export interface MaterialsResponseModel {
  materials: MaterialsData;
}

// ========== КОНФИГУРАЦИЯ ==========

const getApiBaseUrl = (): string => {
  if (typeof window !== 'undefined') {
    const envVar = (window as any).env?.REACT_APP_API_URL;
    if (envVar) return envVar;
  }
  return 'http://localhost:8000/api';
};

const API_CONFIG = {
  baseURL: getApiBaseUrl(),
  timeout: 15000,
  retries: 3,
  retryDelay: 1000
};

// ========== ОШИБКИ ==========

export class MaterialsAPIError extends Error implements APIErrorInterface {
  public status?: number;
  public code?: string;
  public details?: any;

  constructor(
    message: string,
    status?: number,
    code?: string,
    details?: any
  ) {
    super(message);
    this.name = 'MaterialsAPIError';
    this.status = status;
    this.code = code;
    this.details = details;
  }
}

// ========== УТИЛИТЫ ==========

const delay = (ms: number): Promise<void> => 
  new Promise(resolve => setTimeout(resolve, ms));

const retryRequest = async <T>(
  fn: () => Promise<T>,
  retries: number = API_CONFIG.retries,
  retryDelay: number = API_CONFIG.retryDelay
): Promise<T> => {
  try {
    return await fn();
  } catch (error) {
    if (retries > 0 && error instanceof MaterialsAPIError && error.status !== 404) {
      await delay(retryDelay);
      return retryRequest(fn, retries - 1, retryDelay * 1.5);
    }
    throw error;
  }
};

// ========== HTTP КЛИЕНТ ==========

class HTTPClient {
  private baseURL: string;
  private timeout: number;

  constructor(baseURL: string, timeout: number = API_CONFIG.timeout) {
    this.baseURL = baseURL;
    this.timeout = timeout;
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    const contentType = response.headers.get('content-type');
    
    if (!response.ok) {
      let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
      let details = null;
      
      try {
        if (contentType?.includes('application/json')) {
          const errorData = await response.json();
          errorMessage = errorData.detail || errorData.message || errorMessage;
          details = errorData;
        } else {
          errorMessage = await response.text() || errorMessage;
        }
      } catch {
        // Игнорируем ошибки парсинга
      }
      
      throw new MaterialsAPIError(errorMessage, response.status, undefined, details);
    }

    try {
      if (contentType?.includes('application/json')) {
        return await response.json();
      } else {
        const text = await response.text();
        return text as unknown as T;
      }
    } catch (error) {
      throw new MaterialsAPIError('Некорректный формат ответа от сервера');
    }
  }

  private createAbortController(): AbortController {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);
    
    controller.signal.addEventListener('abort', () => clearTimeout(timeoutId));
    
    return controller;
  }

  async get<T>(endpoint: string, params?: Record<string, string | number>): Promise<T> {
    const url = new URL(`${this.baseURL}${endpoint}`);
    
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== null && value !== undefined) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    const controller = this.createAbortController();

    try {
      const response = await fetch(url.toString(), {
        method: 'GET',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        signal: controller.signal,
      });

      return this.handleResponse<T>(response);
    } catch (error: unknown) {
      if ((error as any)?.name === 'AbortError') {
        throw new MaterialsAPIError('Превышено время ожидания запроса', 408);
      }
      throw error;
    }
  }

  async post<T>(endpoint: string, data?: any): Promise<T> {
    const controller = this.createAbortController();

    try {
      const response = await fetch(`${this.baseURL}${endpoint}`, {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: data ? JSON.stringify(data) : undefined,
        signal: controller.signal,
      });

      return this.handleResponse<T>(response);
    } catch (error: unknown) {
      if ((error as any)?.name === 'AbortError') {
        throw new MaterialsAPIError('Превышено время ожидания запроса', 408);
      }
      throw error;
    }
  }
}

// ========== MATERIALS API ==========

export class MaterialsAPI {
  private client: HTTPClient;

  constructor(baseURL?: string) {
    this.client = new HTTPClient(`${baseURL || API_CONFIG.baseURL}/materials`);
  }

  async getMaterialsBySubject(subjectName: string): Promise<MaterialsData> {
    if (!subjectName || typeof subjectName !== 'string') {
      throw new MaterialsAPIError('Название предмета обязательно');
    }

    const normalizedSubject = this.normalizeSubjectName(subjectName);
    
    return retryRequest(async () => {
      try {
        const response = await this.client.get<MaterialsResponseModel>(`/${normalizedSubject}`);
        return response.materials || {};
      } catch (error) {
        if (error instanceof MaterialsAPIError && error.status === 404) {
          return { [normalizedSubject]: { modules: [] } };
        }
        throw error;
      }
    });
  }

  async getSectionMaterials(sectionId: number): Promise<Section> {
    if (!sectionId || sectionId <= 0) {
      throw new MaterialsAPIError('Корректный ID раздела обязателен');
    }

    return retryRequest(async () => {
      return this.client.get<Section>(`/section/${sectionId}`);
    });
  }

  async getTopicDetails(topicId: number): Promise<Topic & { 
    additional_material?: string;
    block_name?: string;
    section_name?: string;
    subject_name?: string;
    number?: number;
    questions_count?: number;
  }> {
    if (!topicId || topicId <= 0) {
      throw new MaterialsAPIError('Корректный ID темы обязателен');
    }

    return retryRequest(async () => {
      return this.client.get(`/topic/${topicId}`);
    });
  }

  async searchMaterials(
    query: string, 
    filters?: { 
      subject_filter?: string;
      section_filter?: number;
      limit?: number;
    }
  ): Promise<SearchResult[]> {
    if (!query || query.trim().length < 2) {
      throw new MaterialsAPIError('Поисковый запрос должен содержать минимум 2 символа');
    }

    const params: Record<string, string | number> = { 
      query: query.trim() 
    };
    
    if (filters?.subject_filter) {
      params.subject_filter = this.normalizeSubjectName(filters.subject_filter);
    }
    if (filters?.section_filter) {
      params.section_filter = filters.section_filter;
    }
    if (filters?.limit && filters.limit > 0) {
      params.limit = Math.min(filters.limit, 100);
    }

    return retryRequest(async () => {
      try {
        return this.client.get<SearchResult[]>('/search', params);
      } catch (error) {
        if (error instanceof MaterialsAPIError && error.status === 404) {
          return [];
        }
        throw error;
      }
    });
  }

  async getStatistics(): Promise<MaterialsStatistics> {
    return retryRequest(async () => {
      return this.client.get<MaterialsStatistics>('/statistics');
    });
  }

  async addFileToSection(
    sectionId: number, 
    fileData: {
      title: string;
      author: string;
      size: string;
      format: string;
      file_type: FileType;
      url?: string;
    }
  ): Promise<{ status: string; message: string; file_id: number }> {
    if (!sectionId || sectionId <= 0) {
      throw new MaterialsAPIError('Корректный ID раздела обязателен');
    }

    this.validateFileData(fileData);

    return retryRequest(async () => {
      return this.client.post(`/section/${sectionId}/files`, fileData);
    });
  }

  async healthCheck(): Promise<{ status: string; service: string; version: string; timestamp: string }> {
    try {
      return await this.client.get('/health');
    } catch (error) {
      throw new MaterialsAPIError('Сервис материалов недоступен');
    }
  }

  private normalizeSubjectName(subjectName: string): string {
    const subject = subjectName.toLowerCase().trim();
    
    if (subject.includes('химия') || subject.includes('chemistry') || subject === 'chem') {
      return 'chemistry';
    }
    
    if (subject.includes('биология') || subject.includes('biology') || subject === 'bio') {
      return 'biology';
    }
    
    return subject;
  }

  private validateFileData(fileData: {
    title: string;
    author: string;
    size: string;
    format: string;
    file_type: string;
  }): void {
    if (!fileData.title?.trim()) {
      throw new MaterialsAPIError('Название файла обязательно');
    }
    if (!fileData.author?.trim()) {
      throw new MaterialsAPIError('Автор файла обязателен');
    }
    if (!fileData.size?.trim()) {
      throw new MaterialsAPIError('Размер файла обязателен');
    }
    if (!Object.values(FILE_TYPES).includes(fileData.file_type as FileType)) {
      throw new MaterialsAPIError('Тип файла должен быть book или test_book');
    }
  }
}

// ========== ЭКЗЕМПЛЯР API ==========

export const materialsAPI = new MaterialsAPI();
export default materialsAPI;